content-api/hierarchy-manager/src/main/scala/org/sunbird/managers/HierarchyManager.scala

```diff
@@ -128,17 +128,21 @@ object HierarchyManager {
 
             val hierarchy = fetchHierarchy(request, rootNode.getIdentifier)
             hierarchy.map(hierarchy => {
-                if (!hierarchy.isEmpty && CollectionUtils.isNotEmpty(hierarchy.getOrDefault("children", "").asInstanceOf[util.ArrayList[java.util.Map[String, AnyRef]]]))
-                    metadata.put("children", hierarchy.getOrDefault("children", new util.ArrayList[java.util.Map[String, AnyRef]]).asInstanceOf[util.ArrayList[java.util.Map[String, AnyRef]]])
-                metadata.put("identifier", request.get("rootId"))
-                if(StringUtils.isNotEmpty(bookmarkId))
-                    metadata = filterBookmarkHierarchy(metadata.get("children").asInstanceOf[util.List[util.Map[String, AnyRef]]], bookmarkId)
-                if (MapUtils.isEmpty(metadata)) {
-                    ResponseHandler.ERROR(ResponseCode.RESOURCE_NOT_FOUND, ResponseCode.RESOURCE_NOT_FOUND.name(), "bookmarkId " + bookmarkId + " does not exist")
-                } else {
-                    ResponseHandler.OK.put("content", metadata)
-                }
-            })
+                val children = hierarchy.getOrDefault("children", new util.ArrayList[java.util.Map[String, AnyRef]]).asInstanceOf[util.ArrayList[java.util.Map[String, AnyRef]]]
+                val leafNodeIds = new util.ArrayList[String]()
+                fetchAllLeafNodes(children, leafNodeIds)
+                getLatestLeafNodes(leafNodeIds).map(leafNodesMap => {
+                    updateLatestLeafNodes(children, leafNodesMap)
+                    metadata.put("children", children)
+                    if(StringUtils.isNotEmpty(bookmarkId))
+                        metadata = filterBookmarkHierarchy(metadata.get("children").asInstanceOf[util.List[util.Map[String, AnyRef]]], bookmarkId)
+                    if (MapUtils.isEmpty(metadata)) {
+                        ResponseHandler.ERROR(ResponseCode.RESOURCE_NOT_FOUND, ResponseCode.RESOURCE_NOT_FOUND.name(), "bookmarkId " + bookmarkId + " does not exist")
+                    } else {
+                        ResponseHandler.OK.put("content", metadata)
+                    }
+                })
+            }).flatMap(f => f)
```

Comment:
Line-151: Implement the latest leafNodes for bookmark hierarchy also.

Reply:
will take it in next pr.